Python/bytecodes.c

```diff
@@ -2693,22 +2685,22 @@ dummy_func(
 
         replaced op(_POP_JUMP_IF_FALSE, (cond -- )) {
             assert(PyStackRef_BoolCheck(cond));
-            int flag = PyStackRef_Is(cond, PyStackRef_False);
+            int flag = PyStackRef_IsFalse(cond);
             DEAD(cond);
             RECORD_BRANCH_TAKEN(this_instr[1].cache, flag);
             JUMPBY(oparg * flag);
         }
 
         replaced op(_POP_JUMP_IF_TRUE, (cond -- )) {
             assert(PyStackRef_BoolCheck(cond));
-            int flag = PyStackRef_Is(cond, PyStackRef_True);
+            int flag = PyStackRef_IsExactly(cond, PyStackRef_True);
```

Comment:
Why do we use `PyStackRef_IsExactly` here (which doesn't mask out the deferred bit) but use `PyStackRef_IsFalse` (which does mask out the deferred bit) in `_POP_JUMP_IF_FALSE` above? Is this the rare case where it's safe?

Reply:
Our codegen ensures that these ops only see True or False. That's often by adding a `TO_BOOL` immediately before, which may be [folded into `COMPARE_OP`](https://github.com/python/cpython/blob/09c240f20c47db126ad7e162df41e5c2596962d4/Python/flowgraph.c#L1829-L1834). The preceding `TO_BOOL`, including in `COMPARE_OP` ensures the canonical representation of `PyStackRef_False` or `PyStackRef_True`.

However, there are two places in `codegen.c` that omit the `TO_BOOL` because they have other reasons to know that the result is exactly a boolean:

https://github.com/python/cpython/blob/09c240f20c47db126ad7e162df41e5c2596962d4/Python/codegen.c#L678-L682

https://github.com/python/cpython/blob/09c240f20c47db126ad7e162df41e5c2596962d4/Python/codegen.c#L5746-L5749

These `COMPARE_OP` here always generated bools, but not always the canonical representation. So we can either:

1) Modify `COMPARE_OP` to ensure the canonical representation like https://github.com/colesbury/cpython/blob/5583ac0c311132e36ef458842e087945898ffdec/Python/bytecodes.c#L2409-L2416
2) Use `PyStackRef_IsFalse` (instead of `PyStackRef_IsExactly`) in the `JUMP_IF_FALSE`
3) Modify the codegen by inserting `TO_BOOL` in those two spots.

Reply:
That makes sense, thanks for the explanation. Since using `PyStackRef_IsExactly` safely is sensitive to code generation changes, I might suggest using it only when we're sure it actually matters for performance, and default to using the variants that mask out the deferred bits everywhere by default since those are always safe. I'd guess that this wouldn't affect the performance improvement of this change much, since it should come from avoiding the tagging in `_PyStackRef_FromPyObjectSteal`. I don't feel super strongly though.

Reply:
I'll switch to using `PyStackRef_IsFalse` and `PyStackRef_IsTrue`.

I'm no longer convinced that `PyStackRef_IsExactly` is actually a performance win (and I didn't see it in measurements). I think we have issues with code generation quality that we'll need to address later. Things like `POP_JUMP_IF_NONE` are composed of `_IS_NONE` and `_POP_JUMP_IF_TRUE` and we pack the intermediate result in a tagged `_PyStackRef`. Clang does a pretty good job of optimizing through it. GCC less so: https://gcc.godbolt.org/z/Ejs8c78qd.